### DISCUSSION TITLE
RXFUNCADD

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#270] [Q&A]
In preparation to call external DLLs, I added a new instruction OPENDLL. As I need the Windows LoadLibrary function, I had to insert #include <windows.h>. I think we need to customise the whole part to the various OSes.  
I have the hope we can write RXFUNCADD as a REXX bif, we will see!

**Comment by Peter-Jacob** [comment]
I am now able to call from CREXX a function from within a DLL. I currently just display a message box, no input/return parms are yet set. The DLL was created with Purebasic, a compilable basic dialect.

**Comment by Peter-Jacob** [comment]
Sending parms to a DLL and returning values works now. I am now able to code an `RXFUNCADD(external_name, library, internal_name)` bif. But I stuck how to define a dynamic parm list in REXX and build it in C

Currently, I have a fixed list containing 3 parameters with int, int string
```
typedef char * (*strSubproc)(int arg1, int arg2, char str1[32]);
typedef int    (*intSubproc)(int arg1, int arg2, char str1[32]);
typedef double (*floatSubproc)(int arg1, int arg2, char str1[32]);
```
I have no idea to make it dynamic based on REXX definitions. 

I am grateful for any advice.

**Reply by rvjansen** [reply]
I am not sure if I understand correctly. I have added an example of how the DES function is called (lib/rxfns/native/des/RxDesTest.rexx) - there was a .gitignore against .cmd, but being OS/2 originally, that was exactly its extension). So I think when we have the signature of the function, we should be able to call it. It is easy to avoid variable argument lists: in this stage, a Rexx external function always needs its interface in a separate program, which can be C and opt for no varargs. Or do I misread this?

**Reply by rvjansen** [reply]
It might be (.. it is certain  ...) that @BobBolch is more knowledgeable on this than I am.

**Comment by BobBolch** [comment]
I am not sure about what you have in mind, but in CMS, an external function "makes itself available" by using CMS system interfaces to declare itself. The function is called by name and all arguments and the result are character strings, which are processed by the caller into numbers or characters. The function package might become available for the duration of a CMS session, or it might be available only when REXX is called by specific applications, like an editor.

**Reply by rvjansen** [reply]
What we have in mind is a rxas instruction that can do a dlopen() for a shared library and load it, so we can store its entry points for calling from crexx. Having done this only on z/OS and OS/2 - indeed z/OS has the interfaces, and in OS/2 it is something between its RxFuncAdd implementation (in the Rexx interpreter) and your RexxSAA compatible calling-the-called program. But I seem to see what Peter is getting at: how to make it so that the routine in the Crexx runtime knows how call your in-between program. I thing we need a standard control block and an argument count.

**Comment by Peter-Jacob** [comment]
Let me try to re-word my question. What we have is an OPENDLL instruction, which allows fetching a DLL and opening a specific entry point (sub procedure), this is straightforward. Now we need to call the entry point with certain input parameters and a return value.  
if it's in C I write:

```
typedef int    (*intSubproc)(int arg1, int arg2, char str1[32]);     // this is the fixed parameter and return definition

hDLL = LoadLibrary("C:/Users/PeterJ/Dropbox/PeterJ/CREXXDLL.dll");   // Load ADDR of DLL
intProc = (intSubproc)GetProcAddress(hDLL, "GetInteger");            // Load entry point, here GetInteger
 int1=intproc(parm1,parm2,parm3);                                   // call the entry point and receive the result integer value
```    
The typedef is the specific "prototype" for this call, it is fixed. We need a variable definition that can be created by a CREXX bif. 

Years ago I played with Reginald REXX a Regina dialect. they used a REXX FUNCDEF function for setting up the DLL call 

```
say FUNCDEF('CHANGE','32,32,str,str,32,32,str',dll-path,'CHANGE')
say FUNCDEF('SEARCH','32,32,str,32,32,32,str,str',dll-path,'SEARCH')

```
meaning the parm-list types are defined (e.g. 32 integer, str string, etc.)

**Reply by rvjansen** [reply]
Yes, although I hope we can limit the os dependent stuff here. Dlopen and acquiring entrypoints works compatible at least over linux, mac and windows (this wasn't the case in 2005 when this was done for ooRexx on Mac). Also, I would not want to deviate from the Rexx standard when not needed.

I think the crexx runtime only calls the interface program, in this case RxDes.dll (.so, .dylib). The functions in there need to all have the same signature:

ULONG RxDesDecrypt(
     PSZ       Name,
     ULONG     Argc,
     RXSTRING  Argv[],
     PSZ       Queuename,
     PRXSTRING Retstr)

ULONG RxDesEncrypt(
     PSZ       Name,
     ULONG     Argc,
     RXSTRING  Argv[],
     PSZ       Queuename,
     PRXSTRING Retstr)

In fact you cannot, in RexxSAA, call a function with a different signature. As these are compiled into the function dll (same goes for function packages, altough more complications there I think) there is only one signature to worry about. 
I think C is loose enough to regard this as a binary signature, that needs to be implemented for the different calling conventions (although we can (and should) prescribe the function interface always to use one particular one).

**Reply by rvjansen** [reply]
I have checked in a rexxsaa.h which works with the RxDes.c source - which after very small changes compiles (even after 29 years , needed some switches to have it accept the ancient C dialect). Am working on Cmake for it.

**Reply by rvjansen** [reply]
Hi Peter, both shared libraries are now in the build - yielding dylibs on macOS and .so’s on Linux. Windows I don’t know - I sincerely hope two .dll’s are coming out.

**Comment by rvjansen** [comment]
So you mentioned Reginald, and that reminded me of Florian's GCI - compatibility with this might serve a larger audience: https://www.rexxla.org/presentations/2004/fgc.pdf - this seems a nice to have, and we have most of the code.

**Reply by Peter-Jacob** [reply]
seems to be a similar approach as in Reginald:
```
say FUNCDEF('CHANGE','32,32,str,str,32,32,str',dll-path,'CHANGE')
say FUNCDEF('SEARCH','32,32,str,32,32,32,str,str',dll-path,'SEARCH')

```
I have to think about how to map this into a generic C parm list

**Comment by rvjansen** [comment]
and what the Reginald approach has got going for it, is that it does not need stems and you could start on that now ...

**Comment by Peter-Jacob** [comment]
A brief update on the DLL support. With the help of MikeG I was able to setup 2 new instructions:
OPENDLL and DLLFUNC. OPENDLL loads a DLL and a contained function.
DLLFUNC calls the loaded function with parameters (to be precise currently just one) and returns the result.
```
/* rexx test DLL bif */
options levelb
DLL="C:\Users\PeterJ\Dropbox\PeterJ\CREXXDLL.dll"
module="GetInteger"
addr=0
rc=0
int1=47

ASSEMBLER opendll addr,DLL,module /* opens a DLL and fetches the addess of a contained function */
say "DLL Func ADDR "rc

ASSEMBLER dllfunc rc,addr,int1         /* calls the function via the returned address and pass one parameter */
say "DLL Func call Result "rc
return
```
The limitation of one is caused by just having 3 instruction parameters, here we can just use the third to pass a parameter. I guess we need something like global variables which can be set by the REXX script and accessed from within an instruction. 

We also need a way to register dynamically a function (name) , that in the subsequent process we can call the DLL function as it was a REXX function. Regina uses `RxFuncAdd(‘rexx-name’,’DLL-Name’, DLL-function-name’)`  

Currently only available for Windows but Mike has shown me a way to make it portable.

**Reply by rvjansen** [reply]
I checked in a Mac/unix version of your opendll instruction (actually the part outside of the ifdef ). Did you check what is in /lib/rxfns/native? There is a rexxsaa.h which has the declarations, and a RxDes example with its native implementation. What is not clear to me is how we can have the 
```
ULONG RxDesEncrypt(
PSZ Name,
ULONG Argc,
RXSTRING Argv[],
PSZ Queuename,
PRXSTRING Retstr)
```
signature implemented in an RXFUNCADD built-in. For this, I think we will need a structure that contains the names of the entrypoints (the example has two, RxDesEncrypt and RxDesDecrypt) and their addresses, and the start address of the parameter structure (which can be the standard CDECL layout of the platform (OS/HW architecture)). We probably need some instructions to marshall the Rexx parameters to their binary equivalents (and we can leave out the queue name until queues are implemented).

**Reply by rvjansen** [reply]
And for clarity, each RXFUNCADD adds one entry point . In OS/2 at least, these were added globally - another Rexx exec in another command shell saw these as already registered, and an RXFUNCDROP made them disappear for all.

**Reply by rvjansen** [reply]
so - thinking aloud further - when rxfuncadd adds one entrypoint a time from a dll to a table in the runtime, the runtime needs to go through that table the moment an external function is called which cannot be found when looking up the linked built-in functions. In the future the rxc compiler might have a role here in assisting the linkage editor/loader to preload the functions that are actually called from the Rexx source.
